File: frontend/src/api/publication.ts
import { apiGet, apiGetBlob, apiHead, apiPost } from './client';
import type { RequirementItemStatus, RequirementItemType } from './item-formation';

/**
 * 模型拒绝为本章节起草初稿（AEP-110 的 status='declined'）。
 *
 * 这不是请求出错：拒绝是模型对「依据不足，照写就是编造」的正当判断，界面要把 `reason` 当
 * 一等回执原样呈现，而不是套用通用报错气泡（通用气泡会在理由前拼上请求地址与 HTTP 状态码，
 * 把真正的理由挤出视野——T20260721 的源头报障正是这个形态）。
 */
export class ManuscriptDeclinedError extends Error {
  readonly reason: string;

  constructor(reason: string) {
    super(reason || 'AI 无法起草该章节初稿');
    this.name = 'ManuscriptDeclinedError';
    this.reason = reason;
  }
}

// ---- SCN-005 发布管理 DTO（与 backend/app/api/schemas.py 同名对齐）----

export type DocumentStatus =
  | 'index_draft'
  | 'index_blocked'
  | 'index_ready'
  | 'markdown_draft'
  | 'markdown_finalized'
  | 'baseline_published';

export type SlotAssetType = 'requirement_item' | 'material' | 'chart' | 'boilerplate';

export type EditImpact =
  | 'doc_expression'
  | 'confirmed_item'
  | 'index_structure'
  | 'no_source_fact'
  | 'other_asset';

export type MarkdownDraftStatus = 'draft' | 'finalized' | 'superseded' | 'awaiting_item_revision';

export type DocxExportStatus =
  | 'converting'
  | 'succeeded'
  | 'failed'
  | 'check_rejected'
  | 'baseline_confirmed'
  | 'manual_fallback';

export interface TemplateSectionRead {
  key: string;
  number: string;
  title: string;
  level: number;
  purpose: string;
  content_types: string[];
  required: boolean;
  repeatable: boolean;
  missing_policy: string;
  boilerplate?: string | null;
  examples?: string[]; // 章节样例（P1 加法字段；复制起草反填/AI 起草少样本）
  /**
   * 撰稿时是否提供「从目录选取」引用标准的入口（T20260721）。
   * 判定在后端单点完成（标题像参考资料类 ∧ 支持人工撰稿），前端只读这个标志——
   * 不要在前端按章节 key 或标题另判一次。
   */
  standards_pickable?: boolean;
}

export interface TemplateDescriptorRead {
  template_ref: string;
  schema_version?: string | null;
  title?: string | null;
  description?: string | null;
  export_binding?: Record<string, unknown> | null;
  sections: TemplateSectionRead[];
  error?: string | null;
}

export interface CandidateItemRead {
  item_ref: string;
  req_no: string;
  expression: string;
  req_type: RequirementItemType;
  status: RequirementItemStatus;
  version_no: string;
}

export interface CandidateMaterialRead {
  material_ref: string;
  source_note: string;
  excerpt: string;
  source_version: number;
}

export interface CandidateChartRead {
  chart_ref: string;
  title: string;
  chart_type: string;
  format: string;
  status: string;
  draft_version: number;
  source_count: number;
  confirmed_at?: string | null;
}

export interface TraceBindingSummaryRead {
  effective: number;
  pre_established: number;
  suspect: number;
}

export interface CandidateAssetsRead {
  items: CandidateItemRead[];
  materials: CandidateMaterialRead[];
  charts: CandidateChartRead[];
  traces: string[];
  trace_summary?: TraceBindingSummaryRead | null;
  pending_item_count: number;
}

export interface DocIndexEntryRead {
  section_key: string;
  asset_type: SlotAssetType;
  asset_ref?: string | null;
  asset_version?: string;
  order_no?: number;
}

export interface SlotStatusRead {
  section_key: string;
  required: boolean;
  satisfied: boolean;
  filled_count: number;
  missing_reason?: string | null;
  rebuild_entry?: string | null;
}

export interface MissingItemRead {
  section_key: string;
  section_title: string;
  reason: string;
  rebuild_entry: string;
}

export interface MarkdownPatchRead {
  patch_ref: string;
  impact: EditImpact;
  before_text: string;
  after_text: string;
  bound_item_ref?: string | null;
  reflow_item_ref?: string | null;
  status: string;
  note?: string | null;
}

export interface SourceBindingRead {
  start_line: number;
  end_line: number;
  kind: 'heading' | 'boilerplate' | 'item' | 'material' | 'chart';
  section_key: string;
  asset_ref?: string | null;
}

export interface DocumentFragmentRead {
  section_key: string;
  section_number: string;
  section_title: string;
  start_line: number;
  end_line: number;
  markdown: string;
}

export interface AssetFragmentRead {
  project_ref: string;
  asset_type: 'requirement_item' | 'chart';
  asset_ref: string;
  document_ref?: string | null;
  document_title?: string | null;
  document_status?: string | null;
  draft_ref?: string | null;
  draft_version?: number | null;
  draft_status?: MarkdownDraftStatus | null;
  index_version?: number | null;
  in_current_index: boolean;
  baseline_ref?: string | null;
  fragments: DocumentFragmentRead[];
  next_action?: string | null;
}

export interface MarkdownDraftRead {
  draft_ref: string;
  version_no: number;
  index_version: number;
  status: MarkdownDraftStatus;
  can_export: boolean;
  content: string;
  source_bindings: SourceBindingRead[];
  block_reasons: string[];
  patches: MarkdownPatchRead[];
  finalized_by?: string | null;
  finalized_at?: string | null;
}

export interface DocxExportRead {
  export_ref: string;
  draft_ref: string;
  status: DocxExportStatus;
  failure_reason?: string | null;
  manual_fallback: boolean;
  check_note?: string | null;
  file_available: boolean;
  created_at: string;
}

export interface ReleaseBaselineRead {
  baseline_ref: string;
  document_ref: string;
  index_version: number;
  draft_ref: string;
  template_ref: string;
  template_schema_version: string;
  export_ref: string;
  manual_fallback: boolean;
  asset_refs: string[];
  confirmed_by: string;
  confirmed_at: string;
  note?: string | null;
}

export interface RequirementDocumentRead {
  document_ref: string;
  doc_type: string;
  title: string;
  template_ref: string;
  template_schema_version: string;
  coverage_scope?: string | null;
  status: DocumentStatus;
  blocked_reason?: string | null;
  index_version: number;
}

export interface SectionManuscriptRead {
  section_key: string;
  content: string;
  revision_no: number;
  updated_by: string;
  updated_at: string;
}

export interface SaveManuscriptCommand {
  project_ref: string;
  template_ref?: string | null;
  section_key: string;
  content: string;
  operator_ref: string;
}

export interface CandidatePreviewRead {
  asset_type: string;
  asset_ref: string;
  title: string;
  markdown: string;
}

/** 某个可撰稿章节的 AI 起草依据计数（口径同后端起草服务实际喂给模型的输入）。 */
export interface SectionDraftBasisRead {
  section_key: string;
  asset_count: number;
  example_count: number;
}

/** AEP-110 起草结果信封：起草成功与模型拒绝都是 HTTP 200。 */
export interface SectionDraftResultRead {
  status: 'drafted' | 'declined';
  manuscript?: SectionManuscriptRead | null;
  reason?: string | null;
}

export interface PublicationWorkspaceRead {
  project_ref: string;
  document?: RequirementDocumentRead | null;
  template: TemplateDescriptorRead;
  candidates: CandidateAssetsRead;
  manuscripts: SectionManuscriptRead[];
  /** 每个可 AI 起草章节的起草依据计数（零依据章节据此在点击前提示）。 */
  draft_basis?: SectionDraftBasisRead[];
  index_entries: DocIndexEntryRead[];
  slot_status: SlotStatusRead[];
  missing_list: MissingItemRead[];
  markdown?: MarkdownDraftRead | null;
  exports: DocxExportRead[];
  baseline?: ReleaseBaselineRead | null;
  next_action?: string | null;
}

export interface SaveIndexCommand {
  project_ref: string;
  template_ref?: string | null;
  coverage_scope?: string | null;
  entries: DocIndexEntryRead[];
  operator_ref: string;
  idempotency_key: string;
}

export interface SaveIndexResult {
  status: 'index_ready' | 'index_blocked' | 'rejected_precheck';
  document_ref?: string | null;
  index_version?: number | null;
  missing_list: MissingItemRead[];
  blocked_reason?: string | null;
  next_action?: string | null;
}

export interface GenerateMarkdownCommand {
  project_ref: string;
  operator_ref: string;
  idempotency_key: string;
}

export interface MarkdownEditCommand {
  project_ref: string;
  draft_ref: string;
  content: string;
  operator_ref: string;
}

export interface MarkdownEditResult {
  status: string;
  draft_ref: string;
  patches: MarkdownPatchRead[];
  can_finalize: boolean;
  block_reasons: string[];
  pending_item_refs: string[];
  next_action?: string | null;
}

export interface FinalizeMarkdownCommand {
  project_ref: string;
  draft_ref: string;
  confirm_reflow?: boolean;
  operator_ref: string;
  idempotency_key: string;
}

export interface FinalizeMarkdownResult {
  status: 'finalized' | 'pending_item_confirmation' | 'item_revision_reflowed' | 'blocked';
  draft_ref: string;
  pending_items: MarkdownPatchRead[];
  reflowed_item_refs: string[];
  block_reasons: string[];
  next_action?: string | null;
}

export interface StartDocxExportResult {
  status: 'submitted' | 'rejected_precheck';
  export_ref?: string | null;
  agent_run_ref?: string | null;
  next_action?: string | null;
}

export interface ConfirmBaselineResult {
  status: 'confirmed' | 'rejected_precheck';
  baseline_ref?: string | null;
  next_action?: string | null;
}

export interface ItemConfirmResult {
  status: string;
  item_ref: string;
  item_status: RequirementItemStatus;
  next_action?: string | null;
}

function base(projectId: string): string {
  return `/projects/${encodeURIComponent(projectId)}/publication`;
}

export const publicationApi = {
  getWorkspace(projectId: string, templateRef?: string): Promise<PublicationWorkspaceRead> {
    const query = templateRef ? `?template_ref=${encodeURIComponent(templateRef)}` : '';
    return apiGet<PublicationWorkspaceRead>(`${base(projectId)}/workspace${query}`);
  },

  saveIndex(projectId: string, command: SaveIndexCommand): Promise<SaveIndexResult> {
    return apiPost<SaveIndexResult>(`${base(projectId)}/index`, command);
  },

  saveManuscript(projectId: string, command: SaveManuscriptCommand): Promise<SectionManuscriptRead> {
    return apiPost<SectionManuscriptRead>(`${base(projectId)}/manuscripts`, command);
  },

  /**
   * AEP-110：为 authored_text 章节 AI 起草初稿（写撰稿阶段，人工可改可确认）。
   *
   * 后端回的是信封，这里在 api 层拆开：起草成功 resolve 撰稿；模型拒绝 reject 一个
   * `ManuscriptDeclinedError`。之所以拆在这层而不是把信封透传给界面——调用链上游只取
   * `content`，透传信封会逼上游每一层都判一次 status；抛错则原样穿过上游，由撰稿抽屉一处
   * 捕获并渲染一等回执。
   */
  async draftManuscript(
    projectId: string,
    sectionKey: string,
    command: { project_ref: string; template_ref?: string | null; operator_ref: string },
  ): Promise<SectionManuscriptRead> {
    const result = await apiPost<SectionDraftResultRead>(
      `${base(projectId)}/manuscripts/${encodeURIComponent(sectionKey)}/draft`,
      command,
    );
    if (result.status === 'declined' || !result.manuscript) {
      throw new ManuscriptDeclinedError(result.reason ?? '');
    }
    return result.manuscript;
  },

  candidatePreview(
    projectId: string,
    assetType: 'requirement_item' | 'chart' | 'material',
    assetRef: string,
  ): Promise<CandidatePreviewRead> {
    const query = `?asset_type=${encodeURIComponent(assetType)}&asset_ref=${encodeURIComponent(assetRef)}`;
    return apiGet<CandidatePreviewRead>(`${base(projectId)}/candidate-preview${query}`);
  },

  getAssetFragment(
    projectId: string,
    assetType: 'requirement_item' | 'chart',
    assetRef: string,
  ): Promise<AssetFragmentRead> {
    const query = `?asset_type=${encodeURIComponent(assetType)}&asset_ref=${encodeURIComponent(assetRef)}`;
    return apiGet<AssetFragmentRead>(`${base(projectId)}/asset-fragment${query}`);
  },

  generateMarkdown(projectId: string, command: GenerateMarkdownCommand): Promise<MarkdownDraftRead> {
    return apiPost<MarkdownDraftRead>(`${base(projectId)}/markdown/generate`, command);
  },

  recordEdit(projectId: string, command: MarkdownEditCommand): Promise<MarkdownEditResult> {
    return apiPost<MarkdownEditResult>(`${base(projectId)}/markdown/edit`, command);
  },

  finalizeMarkdown(projectId: string, command: FinalizeMarkdownCommand): Promise<FinalizeMarkdownResult> {
    return apiPost<FinalizeMarkdownResult>(`${base(projectId)}/markdown/finalize`, command);
  },

  reopenIndex(projectId: string, operatorRef: string): Promise<RequirementDocumentRead> {
    return apiPost<RequirementDocumentRead>(`${base(projectId)}/markdown/reopen-index`, {
      project_ref: projectId,
      operator_ref: operatorRef,
    });
  },

  confirmItem(projectId: string, itemRef: string, operatorRef: string): Promise<ItemConfirmResult> {
    return apiPost<ItemConfirmResult>(`${base(projectId)}/items/${encodeURIComponent(itemRef)}/confirm`, {
      project_ref: projectId,
      item_ref: itemRef,
      operator_ref: operatorRef,
      idempotency_key: `item-confirm-${itemRef}`,
    });
  },

  startExport(projectId: string, draftRef: string, operatorRef: string, idempotencyKey: string): Promise<StartDocxExportResult> {
    return apiPost<StartDocxExportResult>(`${base(projectId)}/exports`, {
      project_ref: projectId,
      draft_ref: draftRef,
      operator_ref: operatorRef,
      idempotency_key: idempotencyKey,
    });
  },

  reportCheck(projectId: string, exportRef: string, passed: boolean, note: string, operatorRef: string): Promise<DocxExportRead> {
    return apiPost<DocxExportRead>(`${base(projectId)}/exports/${encodeURIComponent(exportRef)}/check`, {
      project_ref: projectId,
      export_ref: exportRef,
      passed,
      note,
      operator_ref: operatorRef,
    });
  },

  registerManualFallback(projectId: string, draftRef: string, reason: string, operatorRef: string, idempotencyKey: string): Promise<DocxExportRead> {
    return apiPost<DocxExportRead>(`${base(projectId)}/exports/manual-fallback`, {
      project_ref: projectId,
      draft_ref: draftRef,
      reason,
      operator_ref: operatorRef,
      idempotency_key: idempotencyKey,
    });
  },

  confirmBaseline(projectId: string, exportRef: string, note: string, operatorRef: string, idempotencyKey: string): Promise<ConfirmBaselineResult> {
    return apiPost<ConfirmBaselineResult>(`${base(projectId)}/exports/${encodeURIComponent(exportRef)}/confirm-baseline`, {
      project_ref: projectId,
      export_ref: exportRef,
      note,
      operator_ref: operatorRef,
      idempotency_key: idempotencyKey,
    });
  },

  exportFileUrl(projectId: string, exportRef: string): string {
    return `/api${base(projectId)}/exports/${encodeURIComponent(exportRef)}/file`;
  },

  /** 取生成好的候选/基线 docx 字节流（供在线预览渲染）；与 exportFileUrl 同源。 */
  fetchExportBlob(projectId: string, exportRef: string): Promise<Blob> {
    return apiGetBlob(`${base(projectId)}/exports/${encodeURIComponent(exportRef)}/file`);
  },

  exportPdfUrl(projectId: string, exportRef: string): string {
    return `/api${base(projectId)}/exports/${encodeURIComponent(exportRef)}/pdf`;
  },

  /** 精确预览探活：HEAD 触发/命中转换缓存，2xx 表示可用（iframe 可直连）；未装 LibreOffice 抛 503。 */
  probeExportPdf(projectId: string, exportRef: string): Promise<void> {
    return apiHead(`${base(projectId)}/exports/${encodeURIComponent(exportRef)}/pdf`);
  },
};

File: backend/app/api/schemas/publication.py
"""发布管理：文档编排服务与导出执行服务。

自 schemas.py 按业务域拆包（命名规范见包入口 __init__.py）。
"""
from __future__ import annotations

from pydantic import BaseModel, Field

from app.domain.enums import (
    RequirementItemStatus,
    RequirementItemType,
)


# ---- SCN-005 发布管理（文档编排服务 / 导出执行服务）----


class TemplateSectionRead(BaseModel):
    """模板章节元数据（模板文件适配器抽取；含槽位与必填规则）。"""

    key: str
    number: str
    title: str
    level: int
    purpose: str
    content_types: list[str] = Field(default_factory=list)
    required: bool
    repeatable: bool
    missing_policy: str
    boilerplate: str | None = None
    examples: list[str] = Field(default_factory=list)  # 章节样例（AI 起草少样本；复制起草反填用）
    # 撰稿时是否提供「从目录选取」引用标准的入口（T20260721）：
    # ＝章节标题看起来是参考资料类（domain/reference_standards.py 单点判定） ∧ 支持人工撰稿。
    # 判定结果由后端算好下发，前端不得散落章节 key 字符串。默认 False：模板登记等不涉及撰稿
    # 的读路径不必关心它。
    standards_pickable: bool = False


class TemplateDescriptorRead(BaseModel):
    """模板描述读取结果（schema 校验通过时 sections 非空；失败时 error 说明原因）。"""

    template_ref: str
    schema_version: str | None = None
    title: str | None = None
    description: str | None = None
    export_binding: dict | None = None  # docx 渲染绑定（模板定制器复制起草回填用）
    sections: list[TemplateSectionRead] = Field(default_factory=list)
    error: str | None = None


class CandidateItemRead(BaseModel):
    """候选资产池·确认态需求条目（候选视图不等于入文档许可）。"""

    item_ref: str
    req_no: str
    expression: str
    req_type: RequirementItemType
    status: RequirementItemStatus
    version_no: str


class CandidateMaterialRead(BaseModel):
    """候选资产池·支撑材料。"""

    material_ref: str
    source_note: str
    excerpt: str
    source_version: int


class CandidateChartRead(BaseModel):
    """候选资产池·受控图表（status=confirmed 才进入候选；候选≠许可）。"""

    chart_ref: str
    title: str
    chart_type: str
    format: str
    status: str
    draft_version: int
    source_count: int
    confirmed_at: str | None = None


class TraceBindingSummaryRead(BaseModel):
    """LDM-013 追溯绑定只读摘要（追溯依据不入文档内容，仅供候选池追溯 tab 展示）。"""

    effective: int = 0
    pre_established: int = 0
    suspect: int = 0


class CandidateAssetsRead(BaseModel):
    """P01-N04 候选资产视图（追溯依据只读摘要，不作为可编排内容）。"""

    items: list[CandidateItemRead] = Field(default_factory=list)
    materials: list[CandidateMaterialRead] = Field(default_factory=list)
    charts: list[CandidateChartRead] = Field(default_factory=list)
    traces: list[str] = Field(default_factory=list)
    trace_summary: TraceBindingSummaryRead | None = None
    pending_item_count: int = 0


class SectionManuscriptRead(BaseModel):
    """LDM-014.章节撰稿读视图（AEP-098：人工撰写内容为第一类正文来源）。"""

    section_key: str
    content: str
    revision_no: int
    updated_by: str
    updated_at: str


class SectionDraftResultRead(BaseModel):
    """AEP-110 起草结果信封（T20260721 改造）：起草成功与模型拒绝都是 HTTP 200。

    模型拒绝起草（如章节零依据，照它的判断编内容就是编造）是**正常业务结果**而非请求错误，
    所以走 status='declined' 带上理由原文，让界面把理由当一等回执呈现；早先把它当 400 抛，
    理由会被接口层拼的 URL 与状态码前缀淹没。

    仍走 400 的两种情况不变：模型服务不可用（真故障）、非人工撰稿章节的预检拒绝（用错入口）。
    """

    status: str  # drafted / declined
    manuscript: SectionManuscriptRead | None = None  # status=drafted 时必有
    reason: str | None = None  # status=declined 时必有：模型给出的拒绝理由原文


class SectionDraftBasisRead(BaseModel):
    """某个可撰稿章节的「AI 起草依据」计数（T20260721）。

    口径与起草服务实际喂给模型的一致：asset_count＝挂在本章节且已确认的需求条目数，
    example_count＝模板给本章节写的样例数。两者都为 0 时模型通常会拒绝起草，界面据此在
    点击前就提示，把拒绝提前到点击之前。
    """

    section_key: str
    asset_count: int = 0
    example_count: int = 0


class SaveManuscriptCommand(BaseModel):
    """保存章节撰稿：仅可撰稿章节（boilerplate/authored_text）；content 空白 = 回落默认文本。"""

    project_ref: str
    template_ref: str | None = None  # 文档未创建时按此建档（缺省内置模板）
    section_key: str
    content: str
    operator_ref: str


class DraftManuscriptCommand(BaseModel):
    """AEP-110：章节撰稿 AI 起草初稿（仅 authored_text 章节；写撰稿阶段，人工可改可清空）。"""

    project_ref: str
    template_ref: str | None = None
    operator_ref: str


class CandidatePreviewRead(BaseModel):
    """候选资产渲染预览（AEP-099）：与生成稿同一确定性渲染器，预览即最终渲染。"""

    asset_type: str  # requirement_item / chart / material
    asset_ref: str
    title: str
    markdown: str


class DocIndexEntryRead(BaseModel):
    """文档内容索引条目（只有引用与位置，无正文）。"""

    section_key: str
    asset_type: str
    asset_ref: str | None = None
    asset_version: str = "1"
    order_no: int = 0


class SlotStatusRead(BaseModel):
    """槽位满足状态（索引编排页左栏）。"""

    section_key: str
    required: bool
    satisfied: bool
    filled_count: int
    missing_reason: str | None = None
    rebuild_entry: str | None = None


class MissingItemRead(BaseModel):
    """缺失清单条目（补建依据，不是文档正文内容）。"""

    section_key: str
    section_title: str
    reason: str
    rebuild_entry: str


class MarkdownPatchRead(BaseModel):
    """预览编辑补丁（未定稿前不是正式资产）。"""

    patch_ref: str
    impact: str  # EditImpact 稳定码
    before_text: str
    after_text: str
    bound_item_ref: str | None = None
    reflow_item_ref: str | None = None
    status: str
    note: str | None = None


class SourceBindingRead(BaseModel):
    """Markdown 行区间 → 源资产绑定（编辑影响识别依据）。"""

    start_line: int
    end_line: int
    kind: str  # heading / boilerplate / item / material / chart
    section_key: str
    asset_ref: str | None = None


class MarkdownDraftRead(BaseModel):
    """Markdown 中间稿/定稿读视图。"""

    draft_ref: str
    version_no: int
    index_version: int
    status: str  # MarkdownDraftStatus 稳定码
    can_export: bool
    content: str
    source_bindings: list[SourceBindingRead] = Field(default_factory=list)
    block_reasons: list[str] = Field(default_factory=list)
    patches: list[MarkdownPatchRead] = Field(default_factory=list)
    finalized_by: str | None = None
    finalized_at: str | None = None


class DocxExportRead(BaseModel):
    """候选 docx 导出件读视图（候选≠发布）。"""

    export_ref: str
    draft_ref: str
    status: str  # DocxExportStatus 稳定码
    failure_reason: str | None = None
    manual_fallback: bool = False
    check_note: str | None = None
    file_available: bool = False
    created_at: str


class ReleaseBaselineRead(BaseModel):
    """发布基线快照（只读复核视图）。"""

    baseline_ref: str
    document_ref: str
    index_version: int
    draft_ref: str
    template_ref: str
    template_schema_version: str
    export_ref: str
    manual_fallback: bool
    asset_refs: list[str] = Field(default_factory=list)
    confirmed_by: str
    confirmed_at: str
    note: str | None = None


class RequirementDocumentRead(BaseModel):
    """LDM-014 需求文档读视图。"""

    document_ref: str
    doc_type: str
    title: str
    template_ref: str
    template_schema_version: str
    coverage_scope: str | None = None
    status: str  # DocumentStatus 稳定码
    blocked_reason: str | None = None
    index_version: int


class DocumentFragmentRead(BaseModel):
    """资产在 Markdown 稿中的绑定片段（行区间切片；追溯预览用，只读）。"""

    section_key: str
    section_number: str
    section_title: str
    start_line: int
    end_line: int
    markdown: str


class AssetFragmentRead(BaseModel):
    """资产 → 文档片段追溯读视图（追溯依据不入 docx 正文；绑定由生成时落库）。"""

    project_ref: str
    asset_type: str  # requirement_item / chart
    asset_ref: str
    document_ref: str | None = None
    document_title: str | None = None
    document_status: str | None = None
    draft_ref: str | None = None
    draft_version: int | None = None
    draft_status: str | None = None  # MarkdownDraftStatus 稳定码
    index_version: int | None = None
    in_current_index: bool = False
    baseline_ref: str | None = None
    fragments: list[DocumentFragmentRead] = Field(default_factory=list)
    next_action: str | None = None


class PublicationWorkspaceRead(BaseModel):
    """发布管理工作台唯一工作区读视图（索引编排页 + 发布主工作台共用）。"""

    project_ref: str
    document: RequirementDocumentRead | None = None
    template: TemplateDescriptorRead
    candidates: CandidateAssetsRead
    manuscripts: list[SectionManuscriptRead] = Field(default_factory=list)
    # 每个可 AI 起草章节的起草依据计数（零依据章节据此在点击前提示；口径同起草服务）
    draft_basis: list[SectionDraftBasisRead] = Field(default_factory=list)
    index_entries: list[DocIndexEntryRead] = Field(default_factory=list)
    slot_status: list[SlotStatusRead] = Field(default_factory=list)
    missing_list: list[MissingItemRead] = Field(default_factory=list)
    markdown: MarkdownDraftRead | None = None
    exports: list[DocxExportRead] = Field(default_factory=list)
    baseline: ReleaseBaselineRead | None = None
    next_action: str | None = None


class SaveIndexCommand(BaseModel):
    """P01 保存文档内容索引（含准入校验；缺必填→受阻+缺失清单）。"""

    project_ref: str
    template_ref: str | None = None  # 缺省用内置 SRS 模板
    coverage_scope: str | None = None
    entries: list[DocIndexEntryRead] = Field(default_factory=list)
    operator_ref: str
    idempotency_key: str


class SaveIndexResult(BaseModel):
    """P01 索引保存结果。"""

    status: str  # index_ready / index_blocked / rejected_precheck
    document_ref: str | None = None
    index_version: int | None = None
    missing_list: list[MissingItemRead] = Field(default_factory=list)
    blocked_reason: str | None = None
    next_action: str | None = None


class GenerateMarkdownCommand(BaseModel):
    """P02 生成/重新生成 Markdown 中间稿。"""

    project_ref: str
    operator_ref: str
    idempotency_key: str


class MarkdownEditCommand(BaseModel):
    """P02 窗口微调：提交编辑后全文，系统 diff 识别编辑影响并记录补丁。"""

    project_ref: str
    draft_ref: str
    content: str
    operator_ref: str


class MarkdownEditResult(BaseModel):
    """P02 编辑影响识别结果（即时预览反馈）。"""

    status: str  # recorded
    draft_ref: str
    patches: list[MarkdownPatchRead] = Field(default_factory=list)
    can_finalize: bool = True
    block_reasons: list[str] = Field(default_factory=list)
    pending_item_refs: list[str] = Field(default_factory=list)  # 待修订确认态条目
    next_action: str | None = None


class FinalizeMarkdownCommand(BaseModel):
    """P02 定稿确认。confirm_reflow=true 表示用户已确认待修订确认态条目清单。"""

    project_ref: str
    draft_ref: str
    confirm_reflow: bool = False
    operator_ref: str
    idempotency_key: str


class FinalizeMarkdownResult(BaseModel):
    """P02 定稿结果。"""

    status: str  # finalized / pending_item_confirmation / item_revision_reflowed / blocked
    draft_ref: str
    pending_items: list[MarkdownPatchRead] = Field(default_factory=list)
    reflowed_item_refs: list[str] = Field(default_factory=list)
    block_reasons: list[str] = Field(default_factory=list)
    next_action: str | None = None


class ReopenIndexCommand(BaseModel):
    """P02→P01 调整索引编排（当前稿标记需重新生成）。"""

    project_ref: str
    operator_ref: str


class StartDocxExportCommand(BaseModel):
    """P03 发起 docx 导出（只能从可导出的 Markdown 定稿版本进入）。"""

    project_ref: str
    draft_ref: str
    operator_ref: str
    idempotency_key: str


class StartDocxExportResult(BaseModel):
    """P03 导出受理结果（转换经 agent_run_ref 追踪；inline 模式立即完成）。"""

    status: str  # submitted / rejected_precheck
    export_ref: str | None = None
    agent_run_ref: str | None = None
    next_action: str | None = None


class ExportCheckCommand(BaseModel):
    """P03 候选 docx 检查结论承接。"""

    project_ref: str
    export_ref: str
    passed: bool
    note: str | None = None
    operator_ref: str


class ManualFallbackCommand(BaseModel):
    """P03 人工降级导出件登记（必须标记，不算系统转换成功）。"""

    project_ref: str
    draft_ref: str
    reason: str
    operator_ref: str
    idempotency_key: str


class ConfirmBaselineCommand(BaseModel):
    """P03 发布基线确认（未经用户确认不得形成基线）。"""

    project_ref: str
    export_ref: str
    note: str | None = None
    operator_ref: str
    idempotency_key: str


class ConfirmBaselineResult(BaseModel):
    """P03 基线确认结果。"""

    status: str  # confirmed / rejected_precheck
    baseline_ref: str | None = None
    next_action: str | None = None


class ItemConfirmCommand(BaseModel):
    """需求条目最小确认门禁（SCN-003 完整评审链另行承接）。"""

    project_ref: str
    item_ref: str
    operator_ref: str
    idempotency_key: str


class ItemConfirmResult(BaseModel):
    """条目确认结果。"""

    status: str  # confirmed / rejected_precheck
    item_ref: str
    item_status: RequirementItemStatus
    next_action: str | None = None
